### PR TITLE
Move to a more explicit memory sharing model

### PR DESCRIPTION
**Pull request by sboeuf**
In anticipation for the huge pages support, the way to share/unshare memory from a TVM is updated with a more explicit model that expects the host to drive the entire operation. Relying on some new TEE-Host functions, the host can invalidate and remove a specific range of memory based on the request from the TVM.

### CODE REVIEW

**Comment by sboeuf** [comment]
This PR depends on https://github.com/rivosinc/sbi-rs/pull/12 where the extended TEE-Host API is defined.
Also, this requires some additional tests in `tellus`/`guestvm` to validate it's functional, but I posted the PR to get some early feedback.

**Comment by sboeuf** [comment]
> Also, this requires some additional tests in `tellus`/`guestvm` to validate it's functional, but I posted the PR to get some early feedback.

I've just updated the tellus VM so that it follows the new model for completing memory sharing and memory unsharing from the guest.

File: src/vm_pages.rs
Comment (abrestic-rivos) [inline src/vm_pages.rs:1246]: need to check that all the pages have been removed first

File: page-tracking/src/page_tracker.rs
Comment (sboeuf) [reply page-tracking/src/page_tracker.rs:364]: If you confirm you're okay replacing the `PageInfo::remove()` behavior with `PageInfo::release()` behavior, then of course I'll replace `remove_page()` with `release_page()`.

File: riscv-page-tables/src/page_table.rs
Comment (sboeuf) [reply riscv-page-tables/src/page_table.rs:1081]: Oh so do you also expect that we should be blocking a fully populated range? I'm asking because then I would use the non-sparse version of `invalidate_range()`.

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:1667]: Got it! That makes sense!

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:1701]: No that's the only place where we use it, so I can merge this into a single helper. And I think this will be the same for `promote`/`demote` functions, all they'll care about is if the page is blocked, doesn't matter if it really is `BlockedShared` or `Blocked`.

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:1730]: Yes I'll improve this.

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:1672]: Out of curiosity, can you elaborate why? Do we expect a shared page to be owned by something else than the host?

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:1748]: Oh so you suggest I should lock the regions at the top of this function with `let regions = self.inner.regions.read()` so that no update can happen before the unmapping is performed?

File: src/vm.rs
Comment (sboeuf) [reply src/vm.rs:524]: Ok I'll try that!
What's the clean way to retrieve the value of the A0 register with this approach? I was using the `active_vcpu` but with your proposal I won't have access to it when needed.

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:458]: Yes I'll remove it.

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:1226]: I kept incrementing the TLB (following what was already there in `share_mem_region_begin()`) because that's the way to tell the calling function `handle_tee_guest_msg()` that we have a blocking ecall. But I guess with the new design we can always assume a `ShareMemory` or `UnshareMemory` call is a blocking ecall. And that would allow to get rid of this logic here.

File: page-tracking/src/page_info.rs
Comment (sboeuf) [reply page-tracking/src/page_info.rs:136]: Done!

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:1672]: Ah I see! thanks for the clarification.

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:1246]: Ah yes of course, given we've cleared the PTE bits through `remove_pages`, we expect every PTE from the range to be marked as `Unused`.

File: src/vm.rs
Comment (sboeuf) [reply src/vm.rs:524]: IIUC, the thing you want to avoid is having A0 being read twice, right?

Could we pass a reference of `FinalizedVm` to `VmCpu::complete_pending_op()`? This way, we would invoke `complete_share_mem_region()` from inside VmCpu.
That being said, I'm not a big fan of this approach because `FinalizedVm` is the superset that contains `VmCpu` and I'm not sure it's clean to pass a reference down to `VmCpu`.

Another approach would be to call `VmCpu::complete_pending_op()` first, and expect it to return the value of the A0 register along with the `pending_op`, so that from `vm.rs`, we have all the information we need to call into `complete_share_mem_region()`. The only thing that is unclear to me with this approach is how we can make the return type of `VmCpu::complete_pending_op()` generic enough so that we wouldn't make it specific to the case we're trying to solve here.

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:1667]: I patched this and then realized the `tellus`/`guestvm` test was broken. Basically what happens is that when the `guestvm` first tries to share some untouched memory, then `tellus` invokes `block_pages()`, but then `salus` only finds `Unused` PTEs from the PT walk.
Do I need to reuse the logic that was present in `share_mem_region_begin()` so that trying to share a range if the page is not mapped would not cause a breaking ecall but instead a resumable one? Or instead, do you think the page should have been mapped by `tellus` beforehand?

File: src/vm.rs
Comment (abrestic-rivos) [reply src/vm.rs:524]: > IIUC, the thing you want to avoid is having A0 being read twice, right?

to be more precise, having the A0 that's being acted upon be different from the one that ends up getting written to the guest's register file. but yes, only reading it once would prevent that.

> That being said, I'm not a big fan of this approach because FinalizedVm is the superset that contains VmCpu and I'm not sure it's clean to pass a reference down to VmCpu.

agreed.

> Actually we can keep the take_pending_op() logic associated with calling VmCpu::complete_pending_op() first. That would simplify what needs to be returned by VmCpu::complete_pending_op() (still need to figure out how to make it generic enough).

i suppose that works too. another thing that comes to mind is using a closure that `complete_pending_op()` invokes on the ecall and the contents of A0. something like:

```
pub fn try_complete_pending_op<F, E>(&mut self, ecall_completion: F) -> Result<(), E>
where
    F: Fn(&SbiMessage, SbiReturnType) -> Result<(), E>
{
    match self.arch.pending_op {
        Some(Mmio(mmio_op)) => {
            // handle this same as before
        }
        Some(Ecall(msg)) => {
            let sbi_ret = ...;
            ecall_completion(&msg, sbi_ret)?;
            self.set_ecall_result(sbi_ret);
        }
        None => (),
    }
    self.arch.pending_op = None;
    Ok(())
}
```

and then call `try_complete_pending_op()` before entering the `active_vcpu.run()` loop in `run_vcpu()`.

File: src/vm_pages.rs
Comment (abrestic-rivos) [reply src/vm_pages.rs:1667]: ah, in the case that the region is totally unpopulated to begin with, the host does not have to do the block/remove sequence since there's nothing to remove. so in that case the host can just put the return value in A0 and run the vCPU again.

Comment (sboeuf) [reply src/vm_pages.rs:1667]: Alright!

File: src/vm.rs
Comment (sboeuf) [reply src/vm.rs:524]: Okay I'll implement this way!

**Comment by sboeuf** [comment]
@abrestic-rivos I think I addressed all of your comments.
BTW, are we planning on merging https://github.com/rivosinc/sbi-rs/pull/12? I'm asking because right now this PR relies on my personal fork of `sbi-rs`.

**Comment by atishp04** [comment]
I haven't looked into the PR in details. As per my understanding, this PR will break the existing setup running TVMs as the share memory design is revamped. I was planning to send the RFC with the existing sharing APIs soon. Can we not merge this until host side implementation is also complete ?

The other option is to wait for the sharing APIs to settle down and complete implementation on host/salus side before sending the RFC.

File: page-tracking/src/page_info.rs
Comment (abrestic-rivos) [inline page-tracking/src/page_info.rs:182]: so there's a bit of a wrinkle here i just realized...

we have two ways to end up with a `SharedBlock` page in `release()`:
1. a TVM that had the page blocked was killed before the page was unblocked or removed
2. a TVM that had the shared page mapped (while it was blocked in another TVM) was killed

the behavior here is correct for case (1). for case (2), we want to leave it as `BlockedShared` but decrement the reference count. seems like we need to know whether or not the page was invalidated (indicating it was blocked) when releasing the page in order to differentiate between the two cases.

File: src/vm_pages.rs
Comment (abrestic-rivos) [inline src/vm_pages.rs:1278]: the `*unshare_mem_region` methods look pretty similar to the `*share_mem_region` ones except the region types are flipped. is it possible to deduplicate these a bit?

File: src/vm.rs
Comment (abrestic-rivos) [inline src/vm.rs:2406]: now that the host is responsible for explicitly blocking & removing the pages, we should remove this logic to block until a specific tlb version is met.

File: src/vm_pages.rs
Comment (abrestic-rivos) [inline src/vm_pages.rs:304]: something to consider, not necessarily for this PR, is to adapt the whole IMSIC bind/unbind/rebind bits to internally use the new block/remove flow. that would allow us to completely get rid of the existing "unassignment" flow.

File: page-tracking/src/page_info.rs
Comment (sboeuf) [reply page-tracking/src/page_info.rs:182]: I see... Should we carry the TVM id through the `BlockedShared()` state? That could help us know what to do. Or did you have another idea about how to do it?

**Comment by sboeuf** [comment]
> I haven't looked into the PR in details. As per my understanding, this PR will break the existing setup running TVMs as the share memory design is revamped. I was planning to send the RFC with the existing sharing APIs soon. Are we planning to merge this PR soon ? We need to point to a working salus branch instead of main in the docs in that case.
> 
> The other option is to wait for the sharing APIs to settle down and complete implementation on host/salus side before sending the RFC. That works for me as well as it may be little odd to change the APIs between between RFC and v1 :)

IMO it'd be best to update the RFC to match the new design, otherwise we might confuse the reviewers upstream, and this might increase the delay in getting the RFC agreed on.

File: page-tracking/src/page_info.rs
Comment (abrestic-rivos) [reply page-tracking/src/page_info.rs:182]: i did think about that, but it grows the `PageInfo` struct by 64-bits. that's a non-trivial amount of overhead considering the struct is repeated for every page of memory in the system.

another option is to have separate, e.g.`release_present()` vs `release_invalidated()`, calls for the two paths.

File: src/vm.rs
Comment (abrestic-rivos) [reply src/vm.rs:2406]: i think we can drop the `Blocked` state entirely now that we handle that logic in `complete_pending_op()`.

File: page-tracking/src/page_info.rs
Comment (sboeuf) [reply page-tracking/src/page_info.rs:182]: >another option is to have separate, e.g.release_present() vs release_invalidated(), calls for the two paths.

Ok I'm gonna try this approach

File: src/vm.rs
Comment (sboeuf) [reply src/vm.rs:2406]: Sounds good

File: page-tracking/src/page_info.rs
Comment (sboeuf) [reply page-tracking/src/page_info.rs:200]: I've added the commit https://github.com/rivosinc/salus/pull/261/commits/f1bdd023ea5d41f82b377909d1724c664ad54e66 to handle this issue, please let me know if this is covering what you had in mind.

File: page-tracking/src/page_info.rs
Comment (abrestic-rivos) [inline page-tracking/src/page_info.rs:169]: nit: i think it'd be slightly cleaner if you had a single release with a `was_invalid: bool` param that was used to determine the behavior in the `BlockedShared` state, rather than having `release()` fall through to `release_invalidated()`.

File: page-tracking/src/page_info.rs
Comment (abrestic-rivos) [reply page-tracking/src/page_info.rs:200]: yes, that looks right to me. squash with this commit?

File: src/vm_pages.rs
Comment (abrestic-rivos) [inline src/vm_pages.rs:1668]: kind of pre-existing and not something that needs to be fixed in this PR, but this logic should probably be in a single `PageTracker` call to avoid bouncing of that lock so much.

File: src/vm_pages.rs
Comment (abrestic-rivos) [inline src/vm_pages.rs:1214]: nit: maybe `do_convert_*` for these helpers instead of `do_share_*`? though i guess "convert" is overloaded...

File: src/vm_cpu.rs
Comment (abrestic-rivos) [inline src/vm_cpu.rs:823]: i think we can totally get rid of the `BlockingEcall` exit cause, right?

File: src/vm.rs
Comment (abrestic-rivos) [inline src/vm.rs:538]: maybe hoist the contents of the closure into it's own function?

**Review by abrestic-rivos — COMMENTED** [review]
just a couple minor comments left, otherwise lgtm -- thanks!

File: src/vm_pages.rs
Comment (sboeuf) [reply src/vm_pages.rs:1214]: Yes that's fine I can use `do_convert_region_*`

File: src/vm.rs
Comment (sboeuf) [reply src/vm.rs:538]: This will be cleaner yes.

File: src/vm_cpu.rs
Comment (sboeuf) [reply src/vm_cpu.rs:823]: We use it for setting up the `pending_op`. Should I replace it with `ForwardedEcall` so that I can fully remove it?

Comment (sboeuf) [reply src/vm_cpu.rs]: As per our offline conversation, I've replaced BlockingEcall with ForwardedEcall.

File: src/vm_pages.rs
Comment (abrestic-rivos) [inline src/vm_pages.rs:566]: per our offline conversation, i think we should allow page faults in `*Removable` regions to continue to be serviceable as if they were the original region type. in other words:
- page faults taken in `*Removable` regions should not be fatal
- host should be allowed to add shared pages to `SharedRemovable` regions, and confidential zero pages to `ConfidentialRemovable`.

Comment (sboeuf) [reply src/vm_pages.rs:566]: Sounds good! What about measured pages?

Comment (sboeuf) [reply src/vm_pages.rs]: Please see the new commit handling this: https://github.com/rivosinc/salus/pull/261/commits/7c997b79e15b00b9a701a501286145f07b947054

Comment (abrestic-rivos) [reply src/vm_pages.rs]: measured pages can only be added prior to tvm finalization, so it's impossible for those to be in a `*Removable` region anyway (which you've handled correctly.)

**Review by abrestic-rivos — APPROVED** [review]
lgtm! @atishp04 / @rajnesh-kanwal how do you want to handle submitting this? do you want to take a stab at getting the kernel side working first?

**Comment by rajnesh-kanwal** [comment]
> lgtm! @atishp04 / @rajnesh-kanwal how do you want to handle submitting this? do you want to take a stab at getting the kernel side working first?

The change overall looks good to me. I don't think there will be problems while supporting it in KVM. Just doing a final review and once done I think we can merge this. I will then start to add host side support in KVM.

File: test-workloads/src/bin/tellus.rs
Comment (rajnesh-kanwal) [inline test-workloads/src/bin/tellus.rs:1000]: Can you just add a small comment for this part explaining why we are saving unshared_spa_range here.
As per my understanding this is so that we can check if we need a fence or not and run the if block on line `810`. Right?
Seemed confusing to me at the start.

**Review by rajnesh-kanwal — APPROVED** [review]
One minor comment otherwise LGTM.

File: test-workloads/src/bin/tellus.rs
Comment (sboeuf) [reply test-workloads/src/bin/tellus.rs:1000]: Yes you understood that well. Of course I can add a comment to make things clearer.

**Pull request merged**
